### PR TITLE
Game loop

### PR DESCRIPTION
**Pull request by itsbaska**
* Create a dumb computer that picks a space at random
* Handle bad input when player is asked if they would like to play again
* Remove comments

Still researching on how to make tests for game loops. Once I figure that out I can erase the code and redo it with TDD in mind.

### CODE REVIEW

**Comment by itsbaska** [comment]
Currently the game-loop is integrated to the minimax. I made some changes according to the comments you left me on the minimax PR. The game should continue looping until the user decides to quit.

File: lib/computer.ex [hunk 1]
```diff
@@ -5,35 +5,35 @@ defmodule Computer do
     Game.mark_spot(game, get_best_move(game, game.current_player))
   end
 
-  def score(game) do
+  def score(game, depth) do
     cond do
-      Game.get_winner(game).winner() == %Computer{}.mark() -> 1
-      Game.get_winner(game).winner() == %Player{}.mark() -> -1
+      Game.get_winner(game).winner() == %Computer{}.mark() -> 10 + depth
+      Game.get_winner(game).winner() == %Player{}.mark() -> -10 + depth
```
Comment (esmevane) [inline lib/computer.ex]: In this case, would it be better to do `-10 - depth`?

File: lib/computer.ex [hunk 2]
```diff
@@ -5,35 +5,35 @@ defmodule Computer do
...
       Game.is_tie?(game) -> 0
     end
   end
 
   def minimax_score("O", scores) do
     scores
-    |> Enum.max_by(fn(x) -> elem(x, 1) end) 
+    |> Enum.max_by(fn({_space, score}) -> score end)
```
Comment (esmevane) [inline lib/computer.ex]: I like this notation more - much clearer 👍

File: lib/game.ex [hunk 3]
```diff
@@ -17,39 +17,54 @@ defmodule Game do
...
+      game.board
+        |> convert_to_rows
+        |> check_line
+        |> Enum.find(&(&1))
+    columns =
+      game.board
+        |> convert_to_columns
+        |> check_line
+        |> Enum.find(&(&1)) 
+    winner = diagonals || rows || columns
```
Comment (esmevane) [inline lib/game.ex]: 🎉

File: lib/game.ex [hunk 4]
```diff
@@ -17,39 +17,54 @@ defmodule Game do
...
+      true -> false
     end
   end
 
-  def change_turn(game, %Player{}), do: %{ game | current_player: %Computer{}}
-  def change_turn(game, %Computer{}), do: %{ game | current_player: %Player{}}
+  def change_turn(game, %Player{}), do: update(game, :current_player, %Computer{})
+  def change_turn(game, %Computer{}), do: update(game, :current_player, %Player{})
 
+  def update(game, key, value) do
```
Comment (esmevane) [inline lib/game.ex]: This is interesting.  Very elm-like.

File: lib/game.ex [hunk 5]
```diff
@@ -17,39 +17,54 @@ defmodule Game do
...
+      :player_1 == key ->
+        %{ game | player_1: value}
+      :player_2 == key ->
+        %{ game | player_2: value}
+      :current_player == key ->
+        %{ game | current_player: value}
+      :winner == key ->
+        %{ game | winner: value}
+      :board == key ->
+        %{ game | board: value}
```
Comment (esmevane) [inline lib/game.ex]: How would you feel about this function if it were pattern matching against arguments?  It seems like in all cases you're checking on `key`.  What are your thoughts on this form?

```elixir
def update(game, :current_player, current_player), do: %{ game | current_player: current_player }
def update(game, :player_1, player_1), do: %{ game | player_1: player_1 }
...etc...
```

File: lib/tic_tac_toe.ex [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      end
+
+    if Game.game_over?(game) do
+      %Message{}.game_over |> Console.print
+      Game.get_winner(game).winner |> win_message
+      Console.play_again |> restart_game
+    else 
+      loop(game) 
+    end
+  end
```
Comment (esmevane) [inline lib/tic_tac_toe.ex]: Just a thought, if you are interested in making this testable, you could tease out the `game =` expression into one function, and then the `if Game.game_over?` section could be a second function, which _accepts a function_.

This notation might not be perfectly valid syntax, but just as an example:

```elixir
def perform_game_turn(game, continue_game) do
  if Game.game_over?(game) do
    ...
  else
    continue_game.(game)
  end
end

def loop(game) do
  game = ...
  perform_game_turn(game, &loop/1)
end
```

File: lib/tic_tac_toe.ex [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  def computer_turn(game) do
+    %Message{}.computer_turn |> Console.print
+    game = Computer.move(game)
+    game.board |> Console.print_board
+    game |> Game.change_turn(game.current_player)
+  end
+
+  def get_user_move(game) do
+    move = Console.get_move
+    case Validator.is_valid_input?(move) do
```
Comment (esmevane) [inline lib/tic_tac_toe.ex]: I forget, did you mention a few weeks back that custom functions couldn't be used as function guards?

**Pull request merged**